[IMPL] Pointer type autogeneration added

curin opened the pull request:
## Table of Contents

* [Code of Conduct](#code-of-conduct)
* [Proposal Implementation](#proposal-implementation)

### Code of Conduct

This project has adopted the code of conduct defined by the [Contributor Covenant](http://contributor-covenant.org/) 
to clarify expected behavior in our community. For more information, see the [.NET Foundation Code of Conduct](http://www.dotnetfoundation.org/code-of-conduct).

## Proposal Implementation

[IMPL] Pointer type auto-generation added
- Pointer types up to a depth of 3 have been generated and added to the project
- Extension methods have been added into Core to support converting to strings from valid types
- Previous pointer tests have been retrofitted for the new types

curin committed:
Merge pull request #1 from Perksey/main

Updating to Head

curin committed:
Merge pull request #2 from Perksey/main

Pull from main

curin committed:
Add autogenerated pointers

-Generation done in pointergen project
-used to regenerate GLFW code
-Passed all relevant tests

Perksey commented on eng/pointergen/Generator.cs:
```diff
@@ -0,0 +1,898 @@
+                File.WriteAllText(Path.Combine(outputDir, name + ".gen.cs"), text);
+            }
+
+            text = GenerateILWeavedPointerType(out name, true, typeChain);
+            File.WriteAllText(Path.Combine(outputDir, name + ".generic.gen.cs"), text);
```
How come IL weaving is used for the single-dimension generic ones?

curin replied:
There is a bit of crossover with all the generics, it was easier to have it IL weave rather than separate it out. The more shared sections the easier it will be to maintain. I could have it not IL weave if you would prefer.

Perksey replied:
I think it's best to address all the extant comments first and then once that's done, we can see if there's an easy way to make the single-dimension method handle generics as well.

Perksey commented on eng/pointergen/Generator.cs:
```diff
@@ -0,0 +1,898 @@
+    /// </summary>
+    /// <param name="ptr"></param>
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    public unsafe static implicit operator byte*({{name}} ptr) => (byte*)Unsafe.AsPointer({{getObjectRef}});
```
```suggestion
    public unsafe static explicit operator byte*({{name}} ptr) => (byte*)Unsafe.AsPointer({{getObjectRef}});
```
Unsafe.AsPointer is unsafe and should be an explicit op.

Perksey commented on eng/pointergen/Generator.cs:
```diff
@@ -0,0 +1,898 @@
+    /// </summary>
+    /// <param name="ptr"></param>
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    public unsafe static implicit operator void*({{name}} ptr) => Unsafe.AsPointer({{getObjectRef}});
```
```suggestion
    public unsafe static explicit operator void*({{name}} ptr) => Unsafe.AsPointer({{getObjectRef}});
```
Unsafe.AsPointer is unsafe and should be an explicit op.

Perksey commented on eng/pointergen/Generator.cs:
```diff
@@ -0,0 +1,898 @@
+    /// </summary>
+    /// <param name="array"></param>
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    public static implicit operator {{name}}(byte[,] array) => MemoryMarshal.CreateSpan(ref array[0, 0], array.Length);
```
Does array.Length get the sum of all dimensions (i.e. the length of the array when flat)?

Khitiara replied:
![image](https://github.com/Perksey/SilkX/assets/3117019/aa41f02d-6afe-45f7-8618-e901a9e66364)
per the docs on the Array type yes

Perksey replied:
Thanks :)

Perksey commented on eng/pointergen/Generator.cs:
```diff
@@ -0,0 +1,898 @@
+    /// </summary>
+    /// <param name="ptr"></param>
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    public static implicit operator void*({{fullName}} ptr) => Unsafe.AsPointer(ref Unsafe.AsRef(in ptr.InteriorRef));
```
```suggestion
    public static explicit operator void*({{fullName}} ptr) => Unsafe.AsPointer(ref Unsafe.AsRef(in ptr.InteriorRef));
```

Perksey commented on eng/pointergen/Generator.cs:
```diff
@@ -0,0 +1,898 @@
+    /// Creates a pointer with the given underlying ref.
+    /// </summary>
+    /// <param name="InteriorRef">The underlying ref.</param>
+    public {{name}}(ref {{rOField}}byte @InteriorRef)
```
Do we want this to be public?

curin replied:
no, but it does need to at least be internal (it is used by generics for their non-generic conversions)

Perksey commented on eng/pointergen/Generator.cs:
```diff
@@ -0,0 +1,898 @@
+                )
+            );
+            IL.Emit.Ldarg_1();
+            IL.Emit.Sizeof<nuint>();
```
```suggestion
            IL.Emit.Sizeof<{types.Length > 1 ? "nuint" : arrayType}>();
```

Perksey commented on eng/pointergen/Generator.cs:
```diff
@@ -0,0 +1,898 @@
+                )
+            );
+            IL.Emit.Ldarg_1();
+            IL.Emit.Sizeof<nuint>();
```
```suggestion
            IL.Emit.Sizeof<{types.Length > 1 ? "nuint" : arrayType}>();
```

Perksey commented on sources/Core/Pointers/MutMut.gen.cs:
```diff
@@ -0,0 +1,285 @@
+        );
+        IL.Emit.Ret();
+        throw IL.Unreachable();
+    }    /// <summary>
+    /// Creates a <see cref="MutMut"/> from a string span.
```
```suggestion
    }

    /// <summary>
    /// Creates a <see cref="MutMut"/> from a string span.
```

Perksey commented on sources/Core/Pointers/MutMutPtr.gen.cs:
```diff
@@ -0,0 +1,285 @@
+    /// <summary>
+    /// The underlying reference
+    /// </summary>
+    public readonly ref MutMut Ref
```
This is frazzling me a little bit. iirc in Discord we decided `MutMutPtr` is read as "a mutable pointer to a mutable pointer to an immutable pointer [to a T]" so should this be `ref MutPtr`?

iirc we also wanted to make this a constant we can tweak just in case we back-track

curin replied:
I think I got this reversed in my head at some point, I should be able to fix it pretty easy.

Perksey replied:
I think there are multiple issues at play here.

The mutability and type names seem to be out of sync. For instance `MutMutPtr` was a `ref readonly MutPtr`, now it's a `ref MutMut`. I think it needs to be `ref MutPtr` (the `ref` is the stand-in for the Mut)

curin replied:
This should be fixed now

Perksey commented on sources/Core/Pointers/MutMut.generic.gen.cs:
```diff
@@ -0,0 +1,332 @@
+    /// <param name="array">The array.</param>
+    /// <returns>The pointer.</returns>
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    public static implicit operator MutMut<T>(string[] array)
```
We should not expose these operators if the innermost dimension is mutable, as strings in C# are not meant to be mutable.

Perksey commented on sources/Core/Pointers/MutPtr.gen.cs:
```diff
@@ -0,0 +1,285 @@
+    /// </summary>
+    public readonly ref readonly byte GetInteriorRef() => ref InteriorRef;
+
+    private readonly ref byte InteriorRef;
```
I think this might need to be a `readonly ref readonly`, could be wrong though.

Perksey commented on sources/SilkTouchX/Program.cs:
```diff
@@ -10,6 +10,8 @@
 using Microsoft.Extensions.Logging.Console;
 using SilkTouchX;
 
+System.Threading.Thread.Sleep(1000);
```
?

curin replied:
Oh I thought I removed that, it was to get the compiler to stop while I was debugging

Khitiara replied:
[The Debugger class](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.debugger?view=net-7.0) exists for future reference, the Launch method pauses and attaches a debugger all at once

Perksey commented on tests/Silk.NET.Core.UnitTests/MutMutTests.cs:
```diff
@@ -1,39 +1,40 @@
 namespace Silk.NET.Core.UnitTests;
 
-public class Ptr2DTests
+public class MutMutTests
```
Can you pull in the latest main (hopefully fixing CI)

Perksey commented on sources/GLFW/glfw3/Glfw.gen.cs:
```diff
@@ -669,7 +669,7 @@ public static unsafe partial class Glfw
     public const int DontCare = -1;
     [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
-    public static Ptr<Cursor> CreateCursor([NativeTypeName("const GLFWimage *")] ConstPtr<Image> image, int xhot, int yhot)
+    public static Mut<Cursor> CreateCursor([NativeTypeName("const GLFWimage *")] Ptr<Image> image, int xhot, int yhot)
```
NICE!

Perksey commented on eng/pointergen/Generator.cs:
```diff
@@ -0,0 +1,898 @@
+    /// <param name="array"></param>
+    [MethodImpl(MethodImplOptions.AggressiveInlining | MethodImplOptions.AggressiveOptimization)]
+    public static implicit operator {{fullName}}({{arrayType}}{{arrayChars}} array)
+        => SilkMarshal.JaggedArrayToPointerArray<{{arrayType}}>(array);
```
We also need to discuss whether we want to expose these operators for mutable array types. If the pointer is mutated by native, the original array won't be modified, rather it'll be the pointer array that's allocated by JaggedArrayToPointerArray/StringArrayToArray. You'd have to read the created pointer back to the original array in order for the modifications to be reflected. It'd be an unfortunate deficiency to only be able to pass arrays of arrays if all but the inner dimension is immutable, but one that may be necessary.

curin committed:
Merge pull request #3 from Perksey/main

Update dotnet.yml

curin committed:
Incorporating Review Feedback into Pointer Type Generation

-Unsafe.AsPointer overloads are now explicit
-String overloads are removed from Mut types
-Fixed outer->inner ordering, and added a toggle

-Removed a debug line left in from previous commit

Perksey commented on eng/pointergen/Generator.cs:
```diff
@@ -0,0 +1,904 @@
+{
+    public static class Generator
+    {
+        const bool OUTERMOST_ON_RIGHT = true;
```
Are we getting our wires crossed on the definition here? I would imply given the current rules that this would be `false` i.e. "pointer1 to pointer2 to byte" the "innermost" (i.e. the most amount of dereferences you need to do to get to that point) is the byte, followed by `pointer2` ergo the innermost is on the right surely?

curin replied:
okay this is to control the order of how MutMutPtr so Ptr is the outermost and Mut is innermost (outermost is on the right, <<Mut>Mut>Ptr). Is that not what we want?

Perksey replied:
I think that's what I originally wanted, but then following [this discussion](https://discord.com/channels/521092042781229087/587346162802229298/1163617967058202624) it was changed to:
```cs
ref int Dereference(MutPtrMut<int> outermost)
{
    ref PtrMut<int> secondOutermost = ref outermost[0];
    ref readonly Mut<int> innermost = ref secondOutermost[0];
    return ref innermost[0];
}
```

Perksey replied:
(the change has not been reflected in the proposal tmk)

curin replied:
fixed

curin committed:
Second Review Commit

-Moved all single dimension pointers to the same function
-fixed up pointer handling
-changed secondary constructor visibility

curin committed:
Switch Outermost direction

-Also fixed needing a function internally

curin committed:
Fix Mutability Issue

curin committed:
Merge pull request #4 from Perksey/main

Update dotnet.yml
